### PR TITLE
fix(short/rust/subjects/04): Use readers/writers instead of reading from stdin/writing to stdout

### PR DESCRIPTION
**Pull request by winstonallo**
All exercises are now functions (`lib.rs`) and take readers/writers as arguments to ensure testability.

### CODE REVIEW

File: rust/subjects/module-04.md [hunk 1]
```diff
@@ -195,30 +207,41 @@ turn-in directory:
...
+Your function must have the following signature:
+```rust
+pub fn duh<W: std::io::Write>(writer: &mut W, basedir: &str);
+```
+
+Example Usage:
+
+```rust
+fn main() {
+    let arg: String = std::env::args().nth(1);
```
Comment (bgebetsb) [inline rust/subjects/module-04.md:229]: `.nth()` returns an Option

**Review by bgebetsb — CHANGES_REQUESTED** [review]
Found 2 small issues in the examples, and I think in ex06 the function should take different argument types for the last two arguments. Other than that everything looks good to me.

One more thing to note is that from what I've seen Clippy will complain when writing things like `Vec<String>` in the function signature if you don't use things that are actually specific to Vectors (and thus writing &[String] would be enough) so maybe we should also write it that way in the signatures.

**Comment by winstonallo** [comment]
Yea this should have been a draft haha, thanks for the feedback!

**Comment by winstonallo** [comment]
> One more thing to note is that from what I've seen Clippy will complain when writing things like `Vec<String>` in the function signature if you don't use things that are actually specific to Vectors (and thus writing &[String] would be enough) so maybe we should also write it that way in the signatures.

Yea it also used to be `&[String]`, but I in my implementations I was always converting to Vec and using `.map` to create a vector of files (for example), which would make an extra conversion step if it was not a Vec already.. But this might be due to me being lazy hahaha, I will look into other implementations.

File: rust/subjects/module-04.md (hunk 1 above)
Comment (winstonallo) [reply rust/subjects/module-04.md:229]: I might have written those mains in the markdown directly 🙈

**Pull request closed without merging**